ctapipe/ml/apply.py

```diff
@@ -0,0 +1,381 @@
+        if len(table) <= self.n_cross_validations:
+            raise ValueError(f"Too few events for {telescope_type}.")
+
+        self.log.info(
+            "Starting cross-validation with %d folds for type %s.",
+            self.n_cross_validations,
+            telescope_type,
+        )
+
+        scores = defaultdict(list)
+        predictions = []
+
+        kfold = self.split_data(
+            n_splits=self.n_cross_validations,
+            shuffle=True,
+            # sklearn does not support numpy's new random API yet
+            random_state=self.rng.integers(0, 2**31 - 1),
+        )
+
+        for fold, (train_indices, test_indices) in enumerate(
+            tqdm(kfold.split(table, table[self.model_component.target]), total=self.n_cross_validations)
+        ):
+            train = table[train_indices]
+            test = table[test_indices]
+            cv_prediction, truth, metrics = self.calculate_metrics(
+                telescope_type, train, test
+            )
+            predictions.append(
+                Table(
+                    {
+                        "tel_type": [str(telescope_type)] * len(truth),
+                        "predictions": cv_prediction,
+                        "truth": truth,
+                    }
+                )
+            )
+
+            for metric, value in metrics.items():
+                scores[metric].append(value)
+
+        for metric, cv_values in scores.items():
+            cv_values = np.array(cv_values)
+            with np.printoptions(precision=4):
+                self.log.info(
+                    "Mean %s score from CV: %.4f ± %.4f",
+                    metric,
+                    cv_values.mean(),
+                    cv_values.std(),
+                )
+        self.cv_predictions[telescope_type] = vstack(predictions)
+
+    def _cross_validate_energy(self, telescope_type, train, test):
+        regressor = self.model_component
+        regressor.model.fit(telescope_type, train)
+        prediction, _ = regressor.model.predict(telescope_type, test)
+        truth = test[regressor.target]
+        r2 = metrics.r2_score(truth, prediction)
+        return prediction, truth, {"R^2": r2}
+
+    def _cross_validate_classification(self, telescope_type, train, test):
+        classifier = self.model_component
+        classifier.model.fit(telescope_type, train)
+        prediction, _ = classifier.model.predict_score(telescope_type, test)
+        truth = np.where(
+            test[classifier.target] == classifier.model.positive_class,
+            1,
+            0,
+        )
+        roc_auc = metrics.roc_auc_score(truth, prediction)
+        return prediction, truth, {"ROC AUC": roc_auc}
+
+    def write(self):
+        Provenance().add_output_file(self.output_path, role="ml-cross-validation")
+        for tel_type, results in self.cv_predictions.items():
+            write_table(results, self.output_path, f"cv_predictions_{tel_type}")
```

Comment:
Should add some metadata to these output tables: they should at the very least have a TAB_VER and some description.